File: src/search_manager.js
const axios = require('axios');
const Logger = require('./logger');

class SearchManager {
  constructor() {
    this.logger = new Logger();
    this.searchUrl = process.env.SEARXNG_URL || 'http://localhost:8888';
    this.keywords = [
      'current', 'today', 'now', 'latest', 'patch', 'update',
      'season', 'event', 'version', 'release', '今', '現在',
      '最新', 'current version', 'this patch', 'today\'s'
    ];
  }

  shouldSearch(userMessage, relevantMemories) {
    // Check if message contains keywords indicating freshness need
    const hasKeyword = this.keywords.some(kw => 
      userMessage.toLowerCase().includes(kw)
    );
    
    // Confidence threshold: if we're not sure, search
    const isUncertain = !relevantMemories || relevantMemories.length === 0;
    
    return hasKeyword || isUncertain;
  }

  async search(query) {
    try {
      this.logger.info(`Searching for: ${query}`);
      
      const response = await axios.get(`${this.searchUrl}/search`, {
        params: {
          q: query,
          format: 'json',
          language: 'ja'
        },
        timeout: process.env.SEARCH_TIMEOUT || 10000
      });

      const results = response.data.results || [];
      
      // Process and verify results
      const verified = results.slice(0, 3).map(result => ({
        title: result.title,
        url: result.url,
        summary: result.content,
        source: this.rateSource(result.url),
        date: this.extractDate(result.content)
      }));

      this.logger.info(`Search completed: ${verified.length} results`);
      return verified;
    } catch (err) {
      this.logger.error('Search error', err);
      return [];
    }
  }

  rateSource(url) {
    if (url.includes('.jp/official') || url.includes('/docs/')) return 'OFFICIAL';
    if (url.includes('wiki')) return 'WIKI';
    if (url.includes('reddit.com')) return 'COMMUNITY';
    return 'OTHER';
  }

  extractDate(content) {
    // Simple date extraction
    const dateMatch = content.match(/\d{4}[-\/]\d{2}[-\/]\d{2}/);
    return dateMatch ? dateMatch[0] : null;
  }
}

module.exports = SearchManager;

File: src/search_manager.py
import requests
import logging
from typing import List, Dict, Optional
from datetime import datetime
import re

logger = logging.getLogger(__name__)

class SearchManager:
    """Web検索管理（SearXNG統合）"""
    
    def __init__(self, base_url: str = "http://localhost:8888", timeout: int = 10):
        self.base_url = base_url
        self.timeout = timeout
        self.available = False
        self._check_connection()
    
    def _check_connection(self) -> bool:
        """SearXNG への接続確認"""
        try:
            response = requests.get(f"{self.base_url}/", timeout=5)
            self.available = response.status_code == 200
            if self.available:
                logger.info("✓ Connected to SearXNG")
            return self.available
        except Exception as e:
            logger.warning(f"SearXNG not available: {e}")
            self.available = False
            return False
    
    def search(self, query: str, language: str = "ja", max_results: int = 10) -> List[Dict]:
        """Web検索を実行"""
        if not self.available:
            logger.warning("Search not available")
            return []
        
        try:
            params = {
                'q': query,
                'format': 'json',
                'language': language,
                'pageno': 1
            }
            
            response = requests.get(
                f"{self.base_url}/search",
                params=params,
                timeout=self.timeout
            )
            
            if response.status_code == 200:
                data = response.json()
                results = []
                for result in data.get('results', [])[:max_results]:
                    results.append({
                        'title': result.get('title', ''),
                        'url': result.get('url', ''),
                        'content': result.get('content', ''),
                        'engine': result.get('engine', []),
                        'timestamp': datetime.now().isoformat()
                    })
                logger.info(f"Search results: {len(results)} items")
                return results
            else:
                logger.error(f"Search error: {response.status_code}")
                return []
        except Exception as e:
            logger.error(f"Error performing search: {e}")
            return []
    
    def rate_source(self, result: Dict) -> Dict[str, any]:
        """情報源の信頼度を評価"""
        url = result.get('url', '')
        
        score = 0.5
        if 'official' in url.lower() or 'github.com' in url:
            score = 0.95
        elif 'wikipedia' in url or 'wiki' in url:
            score = 0.8
        elif 'reddit' in url or 'twitter' in url:
            score = 0.5
        elif 'blog' in url:
            score = 0.4
        
        confidence = 'MEDIUM'
        if score > 0.9:
            confidence = 'HIGH'
        elif score < 0.5:
            confidence = 'LOW'
        
        return {
            'url': url,
            'credibility_score': score,
            'confidence': confidence,
            'verified': score > 0.8
        }

search_manager = SearchManager()
